sycl/source/detail/memory_manager.cpp

```diff
@@ -443,10 +446,17 @@ void copyD2D(SYCLMemObjI *SYCLMemObj, RT::PiMem SrcMem, QueueImplPtr SrcQueue,
 
   if (MemType == detail::SYCLMemObjI::MemObjType::Buffer) {
     if (1 == DimDst && 1 == DimSrc) {
-      Plugin.call<PiApiKind::piEnqueueMemBufferCopy>(
-          Queue, SrcMem, DstMem, SrcXOffBytes, DstXOffBytes,
-          SrcAccessRangeWidthBytes, DepEvents.size(), DepEvents.data(),
-          &OutEvent);
+      if (sameCtxt) {
+        Plugin.call<PiApiKind::piEnqueueMemBufferCopy>(
+            Queue, SrcMem, DstMem, SrcXOffBytes, DstXOffBytes,
+            SrcAccessRangeWidthBytes, DepEvents.size(), DepEvents.data(),
+            &OutEvent);
+      } else {
+        Plugin.call<PiApiKind::piextEnqueueMemBufferCopyPeer>(
+            Queue, SrcMem, QueueDst, DstMem, SrcXOffBytes, DstXOffBytes,
+            SrcAccessRangeWidthBytes, DepEvents.size(), DepEvents.data(),
+            &OutEvent);
+      }
```

Comment:
Do not think it's a good approach. I still think that having the same context as an indicator of that the memory can be copied without going thru host(from SYCL RT POV) is a good abstraction and it's not immediately clear why additional ways to bypass it are needed on this level.

Reply:
The main reason that the Peer to Peer copy is made across contexts is because only a single device per context is currently supported for cuda backend: https://github.com/intel/llvm/issues/4381.

Reply:
I suggest modifying cuda plugin/backend to support multiple devices per context.

Reply:
This feature could still be useful for doing [explicit memory copies](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:explicitmemory) between accessors, as I don't know of any restriction requiring the parent contexts to be the same for both accessors.

Reply:
>> This feature could still be useful for doing explicit memory copies between accessors,

Hm. It's interesting. Currently if accessors are from different context we make sure(move) all of them are in the context which the queue is bound to and then do the copy.

What would happen if we enable P2P copies in SYCL RT in this case? Which context resulting `sycl::event` would be bound to? Queue context or a context where data for src accessors is or where data for dst accessor is?

If both devices were in the same cuda plugin context this copy could be handled in the piEnqueueBufferCopy in the cuda plugin by either doing regular copy or P2P. So, it seems we won't loose performance here.

Reply:
> Currently if accessors are from different context we make sure(move) all of them are in the context which the queue is bound to and then do the copy.

I half-assumed it took this path as well. That is interesting!

> What would happen if we enable P2P copies in SYCL RT in this case? Which context resulting sycl::event would be bound to? Queue context or a context where data for src accessors is or where data for dst accessor is?

I would think it would be on the queue, given the memory copy operation is "enqueued" on that queue. However, thinking about it again I am not sure how you'd even do explicit copy between accessors of different contexts, as device accessors are requested through the command group handle.

Instead, I think the place it would be useful is when migrating accessors between devices in the runtime (is that what is happening here?) I.e. 

```c++
sycl::queue queueA, queueB; // Assume queueA and queueB have disjoint sets of devices.
buffer<int, 1> buff{range<1>{1024}};

queueA.submit([&](handler& cgh) {
  accessor accessorA{buff, cgh, read_write};
  // Do some work here
});

queueB.submit([&](handler& cgh) {
  accessor accessorB{buff, cgh, read_write}; // Should migrate data from queueA parent context to queueB parent context.
  // Do more work here
});
```

> If both devices were in the same cuda plugin context this copy could be handled in the piEnqueueBufferCopy in the cuda plugin by either doing regular copy or P2P. So, it seems we can do at least the same if we go with "cuda plugin exposes several devices in the same context" approach.

I completely agree. The functionality exposed is likely also useful for when support for multiple devices per context is implemented in the CUDA backend. However, like in the example mentioned above it is the user that have (implicitly) ordered a data movement between two different contexts, which the backend cannot change. We should optimize for this case as well, if we can.